### ISSUE TITLE
Example Dockerfile to build HAL tools

### ISSUE THREAD

**Issue by nathanweeks** [ComparativeGenomicsToolkit/hal#108]
HAL tools can be somewhat complicated to build given the dependencies and steps involved. The README.md helps, but it might be worth including a Dockerfile that "formalizes" the build & software environment requirements, and produces a working build for development/testing, as well as the end user.

Here's an example that seems to successfully build/install all targets in hal/bin :

https://github.com/ComparativeGenomicsToolkit/hal/compare/master...harvardinformatics:py27-docker

This was branched from what appears to be the last Python-2.7-compatible HAL commit (sonLib was added as a git submodule and similarly pinned to the last Python-2.7-compatible commit, while jobTree, which is required by hal2assemblyHub.py, doesn't appear to be a moving target anymore, and was just downloaded via wget---though it could be added as submodule). 

Just a suggestion to kick start any discussion.

**Comment by nathanweeks** [comment]
What timing: https://github.com/ComparativeGenomicsToolkit/cactus/issues/113 was closed as I was typing this issue. I'll take a look at what was done for HAL.

**Comment by diekhans** [comment]
thanks @nathanweeks, HAL is included in the cactus docker image.
There is a PR right now to build a full contained cactus docker
that includes all the python code.  We could point people to
this.

JobTree is no longer support, it has evolved into Toil.

**Comment by nathanweeks** [comment]
Thanks! I'll give the full cactus Docker image a spin.

jobTool still appears to be a dependency for hal2assemblyHub.py ; getting that working was one of my initial motivators for building HAL separately.

**Comment by diekhans** [comment]
Sorry, hal2assemblyHub hasn't been maintained.  However, we would love
as PR if you want to work on in.  Having tests would help this.

Personally, it seems like using jobTree/Toil might be overkill for
hal2assemblyHub and multiprocess would do the trick.

Mark

**Comment by glennhickey** [comment]
@diekhans Are assembly hubs not a thing anymore?  Is there a new replacement?  If anyone's still using them we should probably port `hal2assemblyHub.py` to Toil and make sure it's runnable in the cactus docker / local installation.

**Comment by diekhans** [comment]
Assembly hubs are a big thing.  With no tests, the
hal2assemblyHub has gone stale.  Although it might still work if
you pull in jobTree.

It would be great to get it working again.  If @nathanweeks
wants to work on it, I will support him.  Otherwise, it will
end up in the queue of things to do.

**Comment by nathanweeks** [comment]
I support users that want hal2assemblyHub.py to work, so I want it to work :-)

Unfortunately, jobTree appears to be Python 2.7 only, and incompatible with the current Python 3 HAL / cactus software environment.

**Comment by diekhans** [comment]
We should port it to Toil (which is an evolution of jobTree).

It does look like it does some computation complex tasks, such as gcpercent and repeat masker.

Repeatmasker should probably be switch to using dfam due to the licensing issues with repbase.

I am in the process off doing assembly hubs for another project, so this would have been helpful and my head is deep in the process.  I will start a separate ticket for the work.

**Comment by diekhans** [comment]
Ticket is here https://github.com/ComparativeGenomicsToolkit/hal/issues/109

**Comment by nathanweeks** [comment]
FWIW, here's an example Dockerfile for building the python3 HAL with a seemingly-functional hal2assemblyHub.py (including the @glennhickey 's jobTree -> toil port): 

https://github.com/harvardinformatics/hal/commit/a489622025155fba70675108fce2fcd65b2fc990

At least, using the resulting image, hal2assemblyHub.py ran to completion on the small cactus evolverMammals.hal file (a colleague will evaluate on their "real" data set). Note that it doesn't bundle repeatmasker, and so doesn't support the `hal2assemblyHub.py --rmskDir` option (and on a related note, it doesn't bundle enough to make the phyloP components functional).

**Comment by glennhickey** [comment]
Cool.  The [Dockerfile](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/Dockerfile) in Cactus should have all these changes as well.  If there's a need for a stand-alone HAL Dockerfile, you could PR yours here.  

HAL (static) binaries can now be downloaded via Cactus releases.  The first and latest is here: https://github.com/ComparativeGenomicsToolkit/cactus/releases/tag/v1.0.0

It should contain the patched assembly hubs.  There is also Docker image for the release here: quay.io/comparative-genomics-toolkit/cactus:v1.0.0

Any feedback on all this is always appreciated.  I think it'd be a good idea to get phylop into the release and docker images.  It's been flagged here https://github.com/ComparativeGenomicsToolkit/cactus/issues/175

**Comment by nathanweeks** [comment]
Oh, cool---didn't notice the official cactus release! I'll definitely check that out. I've set up "watch releases" notifications for the GitHub cactus repo so I don't miss out in the future!

**Comment by nathanweeks** [comment]
@glennhickey : it looks like the cactus:v1.0.0 container image is missing the hal Python libraries?
```
$ singularity pull docker://quay.io/comparative-genomics-toolkit/cactus:v1.0.0
...
$ singularity shell --containall cactus_v1.0.0.sif 
Singularity> hal2assemblyHub.py 
Traceback (most recent call last):
  File "/usr/local/bin/hal2assemblyHub.py", line 28, in <module>
    from hal.assemblyHub.prepareLodFiles import *
ModuleNotFoundError: No module named 'hal'
Singularity> find / -name assemblyHub 2>/dev/null
Singularity> 
```

**Comment by diekhans** [comment]
ouch, we will fix the missing HAL python packages in cactus.

**Comment by nathanweeks** [comment]
OK, that would be great. hal2assemblyHub.py also requires several UCSC "kent" utilities that aren't in the image (see https://github.com/harvardinformatics/hal/commit/a489622025155fba70675108fce2fcd65b2fc990 for a list I had grepped from the assemblyHub python library; note that the procedure for building from source may differ for ubuntu:18.04)

**Comment by diekhans** [comment]
note that the browser groups distribute  precompiled binaries as well: 

https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/

**Comment by nathanweeks** [comment]
Yeah, before bioconda, that's how I used install needed kent utilities in a more "traditional" interactive compute-server environment. There's no version information encoded in the URL (maybe they've started doing that with https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v369/ and https://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64.v385/ ?), but in my limited experience with a few utilities, they're unlikely to bust compatibility between versions, so just curl'ing/wget'ing the needed executables might be the way to go.

**Comment by nathanweeks** [comment]
FWIW, I've updated the Dockerfile to build phyloP:

https://github.com/harvardinformatics/hal/compare/1fe29b80c06161f45dc74f2624878aa8fd0a8629..fffdac7ea8443b7d3823e44687f04c8e0ad8b3a5

It has not been thoroughly tested (a colleague will try running halTreePhyloP.py). At least the test script that is included fails, though apparently due to a HAL file version mismatch:
```
root@7adb9f1891a8:/hal/phyloP/test# sh test.sh 
hal exception caught: HAL API v2.1 incompatible with format v1.2 HAL file.
hal exception caught: HAL API v2.1 incompatible with format v1.2 HAL file.
ERROR: cannot open blanchette.maf.
```

**Comment by diekhans** [comment]
Thanks; much appreciated ... we are going to combine your work into the Cactus release process.  With
limit resource, we feel that one release that includes Cactus and HAL will be more
tractable.

**Comment by nathanweeks** [comment]
Certainly; this Dockerfile is just to illustrate an approach being used to build the HAL components in isolation for local users who have an immediate need (more tractable for me than trying to simultaneously build/integrate with cactus!). I hope it helps with the cactus Dockerfile.

**Comment by diekhans** [comment]
very helpful and much appreciated!
